File: src/libs/sdk/dotnet/Ums.Sdk.Authorization.Testing/AuthGraphBuilder.cs
using Ums.Sdk.Contracts;

namespace Ums.Sdk.Authorization.Testing;

/// <summary>
/// Fluent builder for fabricating valid <see cref="AuthorizationGraph"/> instances in tests.
/// Designed to make consumer-code unit tests trivial: declare exactly the permissions you need,
/// no JSON, no UMS, no HTTP.
/// </summary>
public sealed class AuthGraphBuilder
{
    private string _schemaVersion = SchemaVersion.Current;
    private string _tenantCode = "TEST_TENANT";
    private string _tenantName = "Test Tenant";
    private string _userEmail = "test.user@example.com";
    private string _systemSuiteCode = "TEST_SUITE";
    private string _roleCode = "TEST_ROLE";
    private string _profileScope = "OrgWide";
    private bool _branchScoped;
    private string _branchCode = "TEST_BRANCH";
    private string _authMethod = "Local";
    private IdpProviderRef? _provider;
    private TimeSpan _validity = TimeSpan.FromHours(1);
    private readonly List<string> _scopes = new();
    private readonly List<(string Resource, string Action, AccessEffect Effect, PermissionSource Source)> _domainPerms = new();
    private readonly List<(string OptionCode, string ActionCode, AccessEffect Effect, PermissionSource Source)> _menuOptions = new();
    private readonly List<(string FlagCode, bool Enabled, string? MatchedCriteriaType)> _flags = new();

    public static AuthGraphBuilder ForTenant(string tenantCode)
    {
        var b = new AuthGraphBuilder { _tenantCode = tenantCode };
        return b;
    }

    public AuthGraphBuilder WithTenantName(string name) { _tenantName = name; return this; }
    public AuthGraphBuilder WithUser(string email) { _userEmail = email; return this; }
    public AuthGraphBuilder WithSystemSuite(string code) { _systemSuiteCode = code; return this; }
    public AuthGraphBuilder WithRole(string code) { _roleCode = code; return this; }
    public AuthGraphBuilder WithBranchScopedProfile(string branchCode)
    {
        _profileScope = "BranchScoped";
        _branchScoped = true;
        _branchCode = branchCode;
        return this;
    }
    public AuthGraphBuilder WithSchemaVersion(string version) { _schemaVersion = version; return this; }
    public AuthGraphBuilder WithValidity(TimeSpan duration) { _validity = duration; return this; }
    public AuthGraphBuilder WithIdpAuth(string providerName, string providerCode, string strategy)
    {
        _authMethod = "IDP";
        _provider = new IdpProviderRef(providerName, providerCode, strategy);
        return this;
    }
    public AuthGraphBuilder WithScope(string scope) { _scopes.Add(scope); return this; }
    public AuthGraphBuilder WithDeny(string scope)
    {
        var (r, a) = ParseScope(scope);
        if (r is null || a is null)
            throw new ArgumentException($"Invalid scope format '{scope}'. Expected 'RESOURCE.ACTION'.", nameof(scope));
        _domainPerms.Add((r, a, AccessEffect.Deny, PermissionSource.Override));
        return this;
    }
    public AuthGraphBuilder WithDomainPermission(string resourceCode, string actionCode,
        AccessEffect effect = AccessEffect.Allow, PermissionSource source = PermissionSource.Template)
    {
        _domainPerms.Add((resourceCode, actionCode, effect, source));
        if (effect == AccessEffect.Allow) _scopes.Add($"{resourceCode}.{actionCode}");
        return this;
    }
    public AuthGraphBuilder WithMenuOption(string optionCode, string actionCode,
        AccessEffect effect = AccessEffect.Allow, PermissionSource source = PermissionSource.Template)
    {
        _menuOptions.Add((optionCode, actionCode, effect, source));
        if (effect == AccessEffect.Allow) _scopes.Add($"{optionCode}.{actionCode}");
        return this;
    }
    public AuthGraphBuilder WithFeatureFlag(string flagCode, bool enabled = true, string? matchedCriteriaType = null)
    {
        _flags.Add((flagCode, enabled, enabled ? (matchedCriteriaType ?? "TenantId") : null));
        return this;
    }

    public AuthorizationGraph Build()
    {
        var now = DateTimeOffset.UtcNow;
        var validUntil = now.Add(_validity);

        var user = new UserSummary(
            Guid.NewGuid(),
            _userEmail,
            _userEmail.Split('@')[0],
            _userEmail,
            "ACTIVE");

        var tenant = new TenantSummary(Guid.NewGuid(), _tenantCode, _tenantName, "ACTIVE");
        var suite = new SystemSuiteSummary(Guid.NewGuid(), _systemSuiteCode, _systemSuiteCode, "PUBLISHED");
        var role = new RoleSummary(Guid.NewGuid(), _roleCode, _roleCode, 1, null);
        var profile = new ProfileSummary(Guid.NewGuid(), _profileScope, true);
        BranchSummary? branch = _branchScoped
            ? new BranchSummary(Guid.NewGuid(), _branchCode, _branchCode)
            : null;

        var context = new PrincipalContext(user, tenant, suite, role, profile, branch);
        var authentication = new AuthenticationMetadata(_authMethod, _provider, false, now, validUntil);

        var actions = new List<ActionRef>();
        var menuAccess = BuildMenuAccess(actions);
        var domainPermissions = BuildDomainPermissions(actions);
        var featureFlags = _flags
            .Select(f => new FeatureFlagState(f.FlagCode, suite.Id, f.Enabled, f.MatchedCriteriaType))
            .ToList();

        var effectiveConfig = new EffectiveConfig(60, 5, 12, true, 3600000, _authMethod == "IDP");

        return new AuthorizationGraph(
            _schemaVersion,
            context,
            authentication,
            actions,
            menuAccess,
            domainPermissions,
            featureFlags,
            effectiveConfig,
            _scopes.Distinct(StringComparer.Ordinal).ToList(),
            now,
            validUntil);
    }

    public AuthorizationGraph BuildExpired()
    {
        var graph = Build();
        var past = DateTimeOffset.UtcNow.AddYears(-5);
        return graph with { GeneratedAt = past, ValidUntil = past.AddHours(1) };
    }

    private List<MenuModule> BuildMenuAccess(List<ActionRef> actions)
    {
        if (_menuOptions.Count == 0) return new();

        var moduleId = Guid.NewGuid();
        var menuId = Guid.NewGuid();
        var subId = Guid.NewGuid();
        var options = _menuOptions.Select(m =>
        {
            var existing = actions.FirstOrDefault(a => a.Code == m.ActionCode);
            var actionRef = existing ?? new ActionRef(Guid.NewGuid(), m.ActionCode, m.ActionCode);
            if (existing is null) actions.Add(actionRef);
            return new MenuOption(Guid.NewGuid(), m.OptionCode, m.OptionCode, m.ActionCode, m.Effect, m.Source);
        }).ToList();

        var submenu = new SubMenu(subId, "TEST_SUB", "Test SubMenu", 1, options);
        var menu = new Menu(menuId, "TEST_MENU", "Test Menu", 1, new[] { submenu });
        var module = new ModuleSummary(moduleId, "TEST_MODULE", "Test Module", 1, "PUBLISHED");
        return new() { new MenuModule(module, new[] { menu }) };
    }

    private List<DomainResourcePermissions> BuildDomainPermissions(List<ActionRef> actions)
    {
        if (_domainPerms.Count == 0) return new();

        return _domainPerms
            .GroupBy(d => d.Resource)
            .Select(grp =>
            {
                var resource = new DomainResource(Guid.NewGuid(), "Aggregate", grp.Key, grp.Key, null);
                var resolutions = grp.Select(d =>
                {
                    var existing = actions.FirstOrDefault(a => a.Code == d.Action);
                    var actionRef = existing ?? new ActionRef(Guid.NewGuid(), d.Action, d.Action);
                    if (existing is null) actions.Add(actionRef);
                    return new DomainActionResolution(actionRef.Id, d.Action, d.Action, d.Effect, d.Source);
                }).ToList();
                return new DomainResourcePermissions(resource, resolutions);
            })
            .ToList();
    }

    private static (string? Resource, string? Action) ParseScope(string scope)
    {
        var idx = scope.IndexOf('.');
        if (idx <= 0 || idx == scope.Length - 1) return (null, null);
        return (scope[..idx], scope[(idx + 1)..]);
    }
}

File: src/libs/sdk/typescript/sdk-testing/src/auth-graph-builder.ts
import type {
  AccessEffect,
  ActionRef,
  AuthorizationGraph,
  DomainResourcePermissions,
  IdpProviderRef,
  MenuModule,
  PermissionSource
} from '@ums/sdk-contracts';
import { SchemaVersion } from '@ums/sdk-contracts';

type FlagCriteria =
  | 'TenantId'
  | 'BranchId'
  | 'UserProfileId'
  | 'RoleCode'
  | 'Environment'
  | 'DateRange'
  | 'PercentageHash'
  | 'CustomRule';

interface MenuOptionSpec {
  optionCode: string;
  actionCode: string;
  effect: AccessEffect;
  source: PermissionSource;
}

interface DomainPermSpec {
  resource: string;
  action: string;
  effect: AccessEffect;
  source: PermissionSource;
}

interface FlagSpec {
  flagCode: string;
  enabled: boolean;
  matchedCriteriaType: FlagCriteria | null;
}

/**
 * Fluent builder for fabricating valid {@link AuthorizationGraph} instances in tests.
 * Same surface as the .NET `AuthGraphBuilder` — produces graphs without JSON parsing,
 * HTTP, or UMS connectivity.
 */
export class AuthGraphBuilder {
  private schemaVersion: string = SchemaVersion.Current;
  private tenantCode: string = 'TEST_TENANT';
  private tenantName: string = 'Test Tenant';
  private userEmail: string = 'test.user@example.com';
  private systemSuiteCode: string = 'TEST_SUITE';
  private roleCode: string = 'TEST_ROLE';
  private profileScope: 'OrgWide' | 'BranchScoped' = 'OrgWide';
  private branchCode: string | null = null;
  private authMethod: 'Local' | 'IDP' = 'Local';
  private provider: IdpProviderRef | null = null;
  private validityMs: number = 60 * 60 * 1000;
  private readonly scopes: string[] = [];
  private readonly menuOptions: MenuOptionSpec[] = [];
  private readonly domainPerms: DomainPermSpec[] = [];
  private readonly flags: FlagSpec[] = [];

  static forTenant(code: string): AuthGraphBuilder {
    const b = new AuthGraphBuilder();
    b.tenantCode = code;
    return b;
  }

  withTenantName(name: string): this {
    this.tenantName = name;
    return this;
  }

  withUser(email: string): this {
    this.userEmail = email;
    return this;
  }

  withSystemSuite(code: string): this {
    this.systemSuiteCode = code;
    return this;
  }

  withRole(code: string): this {
    this.roleCode = code;
    return this;
  }

  withBranchScopedProfile(branchCode: string): this {
    this.profileScope = 'BranchScoped';
    this.branchCode = branchCode;
    return this;
  }

  withSchemaVersion(version: string): this {
    this.schemaVersion = version;
    return this;
  }

  withValidity(ms: number): this {
    this.validityMs = ms;
    return this;
  }

  withIdpAuth(providerName: string, providerCode: string, strategy: string): this {
    this.authMethod = 'IDP';
    this.provider = { name: providerName, code: providerCode, strategy };
    return this;
  }

  withScope(scope: string): this {
    this.scopes.push(scope);
    return this;
  }

  withDeny(scope: string): this {
    const idx = scope.indexOf('.');
    if (idx <= 0 || idx === scope.length - 1) {
      throw new Error(`Invalid scope format '${scope}'. Expected 'RESOURCE.ACTION'.`);
    }
    this.domainPerms.push({
      resource: scope.substring(0, idx),
      action: scope.substring(idx + 1),
      effect: 'Deny',
      source: 'Override'
    });
    return this;
  }

  withDomainPermission(
    resourceCode: string,
    actionCode: string,
    effect: AccessEffect = 'Allow',
    source: PermissionSource = 'Template'
  ): this {
    this.domainPerms.push({ resource: resourceCode, action: actionCode, effect, source });
    if (effect === 'Allow') this.scopes.push(`${resourceCode}.${actionCode}`);
    return this;
  }

  withMenuOption(
    optionCode: string,
    actionCode: string,
    effect: AccessEffect = 'Allow',
    source: PermissionSource = 'Template'
  ): this {
    this.menuOptions.push({ optionCode, actionCode, effect, source });
    if (effect === 'Allow') this.scopes.push(`${optionCode}.${actionCode}`);
    return this;
  }

  withFeatureFlag(flagCode: string, enabled: boolean = true, matchedCriteriaType?: FlagCriteria): this {
    this.flags.push({
      flagCode,
      enabled,
      matchedCriteriaType: enabled ? matchedCriteriaType ?? 'TenantId' : null
    });
    return this;
  }

  build(): AuthorizationGraph {
    const now = new Date();
    const validUntil = new Date(now.getTime() + this.validityMs);

    const actions: ActionRef[] = [];
    const menuAccess = this.buildMenuAccess(actions);
    const domainPermissions = this.buildDomainPermissions(actions);

    return {
      schemaVersion: this.schemaVersion,
      context: {
        user: {
          id: randomUuid(),
          email: this.userEmail,
          username: this.userEmail.split('@')[0] ?? this.userEmail,
          displayName: this.userEmail,
          status: 'ACTIVE'
        },
        tenant: { id: randomUuid(), code: this.tenantCode, name: this.tenantName, status: 'ACTIVE' },
        systemSuite: { id: randomUuid(), code: this.systemSuiteCode, name: this.systemSuiteCode, status: 'PUBLISHED' },
        role: { id: randomUuid(), code: this.roleCode, name: this.roleCode, hierarchyLevel: 1, parentRoleId: null },
        profile: { id: randomUuid(), scope: this.profileScope, isActive: true },
        branch:
          this.profileScope === 'BranchScoped' && this.branchCode !== null
            ? { id: randomUuid(), code: this.branchCode, name: this.branchCode }
            : null
      },
      authentication: {
        method: this.authMethod,
        provider: this.provider,
        mfaRequired: false,
        issuedAt: now.toISOString(),
        sessionExpiresAt: validUntil.toISOString()
      },
      actions,
      menuAccess,
      domainPermissions,
      featureFlags: this.flags.map((f) => ({
        flagCode: f.flagCode,
        systemSuiteId: randomUuid(),
        isEnabled: f.enabled,
        matchedCriteriaType: f.matchedCriteriaType
      })),
      effectiveConfig: {
        sessionTimeoutMinutes: 60,
        maxLoginAttempts: 5,
        minPasswordLength: 12,
        mfaRequiredForAdmin: true,
        accessTokenDurationMs: 3600000,
        authUseExternalIdp: this.authMethod === 'IDP'
      },
      scopes: Array.from(new Set(this.scopes)),
      generatedAt: now.toISOString(),
      validUntil: validUntil.toISOString()
    };
  }

  buildExpired(): AuthorizationGraph {
    const graph = this.build();
    const past = new Date(Date.now() - 5 * 365 * 24 * 60 * 60 * 1000);
    return {
      ...graph,
      generatedAt: past.toISOString(),
      validUntil: new Date(past.getTime() + 60 * 60 * 1000).toISOString()
    };
  }

  private buildMenuAccess(actions: ActionRef[]): MenuModule[] {
    if (this.menuOptions.length === 0) return [];
    const options = this.menuOptions.map((m) => {
      const existing = actions.find((a) => a.code === m.actionCode);
      const actionRef = existing ?? { id: randomUuid(), code: m.actionCode, name: m.actionCode };
      if (!existing) actions.push(actionRef);
      return {
        id: randomUuid(),
        code: m.optionCode,
        label: m.optionCode,
        actionCode: m.actionCode,
        effect: m.effect,
        source: m.source
      };
    });
    return [
      {
        module: {
          id: randomUuid(),
          code: 'TEST_MODULE',
          name: 'Test Module',
          sortOrder: 1,
          status: 'PUBLISHED'
        },
        menus: [
          {
            id: randomUuid(),
            code: 'TEST_MENU',
            label: 'Test Menu',
            sortOrder: 1,
            subMenus: [
              {
                id: randomUuid(),
                code: 'TEST_SUB',
                label: 'Test SubMenu',
                sortOrder: 1,
                options
              }
            ]
          }
        ]
      }
    ];
  }

  private buildDomainPermissions(actions: ActionRef[]): DomainResourcePermissions[] {
    if (this.domainPerms.length === 0) return [];
    const byResource = new Map<string, DomainPermSpec[]>();
    for (const dp of this.domainPerms) {
      const list = byResource.get(dp.resource) ?? [];
      list.push(dp);
      byResource.set(dp.resource, list);
    }
    const result: DomainResourcePermissions[] = [];
    for (const [resourceCode, perms] of byResource) {
      const resolutions = perms.map((p) => {
        const existing = actions.find((a) => a.code === p.action);
        const actionRef = existing ?? { id: randomUuid(), code: p.action, name: p.action };
        if (!existing) actions.push(actionRef);
        return {
          actionId: actionRef.id,
          actionCode: p.action,
          actionName: p.action,
          effect: p.effect,
          source: p.source
        };
      });
      result.push({
        resource: {
          id: randomUuid(),
          type: 'Aggregate',
          code: resourceCode,
          name: resourceCode,
          moduleId: null
        },
        actions: resolutions
      });
    }
    return result;
  }
}

function randomUuid(): string {
  // RFC 4122 v4-ish; cryptographically random when crypto.randomUUID is available.
  const c = globalThis.crypto as { randomUUID?: () => string } | undefined;
  if (c?.randomUUID) return c.randomUUID();
  // Fallback for older environments.
  return 'xxxxxxxx-xxxx-4xxx-8xxx-xxxxxxxxxxxx'.replace(/[xy]/g, (ch) => {
    const r = Math.floor(Math.random() * 16);
    const v = ch === 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}
